File: bms/go/geo/geo_s2_Cap_InteriorContainsPoint/geo_s2_Cap_InteriorContainsPoint.go
package test

import (
	"math"
)

const (
	geo_s1_StraightChordAngle = geo_s1_ChordAngle(4)
)

func (c geo_s2_Cap) InteriorContainsPoint(p geo_s2_Point) bool {
	return c.IsFull() || geo_s2_ChordAngleBetweenPoints(c.center, p) < c.radius
}

func (c geo_s2_Cap) IsFull() bool {
	return c.radius == geo_s1_StraightChordAngle
}

func geo_s2_ChordAngleBetweenPoints(x, y geo_s2_Point) geo_s1_ChordAngle {
	return geo_s1_ChordAngle(math.Min(4.0, x.Sub(y.geo_r3_Vector).Norm2()))
}

func (v geo_r3_Vector) Sub(ov geo_r3_Vector) geo_r3_Vector {
	return geo_r3_Vector{v.X - ov.X, v.Y - ov.Y, v.Z - ov.Z}
}

func (v geo_r3_Vector) Norm2() float64 { return v.Dot(v) }

func (v geo_r3_Vector) Dot(ov geo_r3_Vector) float64 {
	return float64(v.X*ov.X) + float64(v.Y*ov.Y) + float64(v.Z*ov.Z)
}

type geo_s2_Cap struct {
	center geo_s2_Point
	radius geo_s1_ChordAngle
}

type geo_s2_Point struct {
	geo_r3_Vector
}

type geo_r3_Vector struct {
	X, Y, Z float64
}

type geo_s1_ChordAngle float64

File: translations_only/fail/go/geo/geo_s2_Cap_InteriorContainsPoint/gpt4/Restart/geo_s2_Cap_InteriorContainsPoint.rs

struct GeoR3Vector {
    x: f64,
    y: f64,
    z: f64,
}

impl GeoR3Vector {
    fn sub(&self, ov: &GeoR3Vector) -> GeoR3Vector {
        GeoR3Vector {
            x: self.x - ov.x,
            y: self.y - ov.y,
            z: self.z - ov.z,
        }
    }

    fn norm2(&self) -> f64 {
        self.dot(self)
    }

    fn dot(&self, ov: &GeoR3Vector) -> f64 {
        self.x * ov.x + self.y * ov.y + self.z * ov.z
    }
}

struct GeoS2Point {
    geo_r3_vector: GeoR3Vector,
}

struct GeoS1ChordAngle(f64);

struct GeoS2Cap {
    center: GeoS2Point,
    radius: GeoS1ChordAngle,
}

const GEO_S1_STRAIGHT_CHORD_ANGLE: GeoS1ChordAngle = GeoS1ChordAngle(4.0);

fn geo_s2_chord_angle_between_points(x: &GeoS2Point, y: &GeoS2Point) -> GeoS1ChordAngle {
    GeoS1ChordAngle(f64::min(4.0, x.geo_r3_vector.sub(&y.geo_r3_vector).norm2()))
}

fn interior_contains_point(c: &GeoS2Cap, p: &GeoS2Point) -> bool {
    is_full(c) || geo_s2_chord_angle_between_points(&c.center, p).0 < c.radius.0
}

fn is_full(c: &GeoS2Cap) -> bool {
    c.radius.0 == GEO_S1_STRAIGHT_CHORD_ANGLE.0
}
